### ISSUE TITLE
No error despite of multiple applicable methods in scope

### ISSUE THREAD

**Issue by michas2** [rust-lang/rust#26080]
Given the following code:

```rust
extern crate mio;
use mio::buf::RingBuf;
use mio::buf::Buf;
use std::io::Read;

fn main() {
    let buf = RingBuf::new(10);
    let bytes = buf.bytes();
    println!("{:?}", bytes);
}
```
`buf` is of type `RingBuf`. `RingBuf` does not provide `.bytes()`, but it implements both `Buf` and `Read`, which both provide a `.bytes()` implementation.

According to https://doc.rust-lang.org/book/ufcs.html the compiler should complain. But it simply chooses to take the implementation of `Read`, which return the "wrong" result.

(Without the "use Read" line rust chooses the implementation of `Buf`, which return the "correct" type.)

**Comment by bluss** [comment]
Reproduction without `mio` follows: Comment out the first line with `use Read` and the return type of `.bytes()` switches. No ambiguities are reported.

```rust
use std::io::Read;

struct RingBuf;

impl std::io::Read for RingBuf {
    fn read(&mut self, b: &mut [u8]) -> std::io::Result<usize> {
        panic!()
    }
}

trait Test {
    fn bytes(&self) -> &[u8] { &[] }
}
impl Test for RingBuf { }

fn main() {
    let () = RingBuf.bytes();
}
```

**Comment by arielb1** [comment]
UFCS is not used here (that would be `<RingBuf>::bytes`). Method lookup tries autoderefs/autorefs  in the order `0 autoderef -> 0 autoderef + autoref -> 1 autoderef -> 1 autoderef + autoref -> ...`. This is required for e.g. `rc.clone()` (but, of course, not for `<Rc>::clone`) to work.

**Comment by bluss** [comment]
@michas2 It may be that it's a known "feature" of method look up priorities, yours and my example both use the fact that the methods use `self` vs. `&self` and thus the `self` is preferred.

It's confusing but perhaps not a bug like I thought!

As @arielb1 says, using the `<Typename>::bytes(...)` syntax will produce the ambiguous methods error message.

**Comment by alexcrichton** [comment]
triage: I-nominated

**Comment by pnkfelix** [comment]
triage: still an issue.

**Comment by steveklabnik** [comment]
Triage: still reproduces, not something that I think we're likely to change?

**Comment by Spoonbender** [comment]
Triage: still reproduces

Do we still consider this a bug?
